checker/src/org/checkerframework/checker/lock/LockStore.java

```diff
@@ -62,36 +70,66 @@ public LockStore leastUpperBound(LockStore other) {
      * This is only done for @LockPossiblyHeld. This is not sound for other type qualifiers.
      */
     public void insertLockPossiblyHeld(FlowExpressions.Receiver r) {
-        CFValue value =
-                analysis.createSingleAnnotationValue(atypeFactory.LOCKPOSSIBLYHELD, r.getType());
-        assert value != null;
-
         if (r.containsUnknown()) {
             // Expressions containing unknown expressions are not stored.
             return;
         }
         if (r instanceof FlowExpressions.LocalVariable) {
             FlowExpressions.LocalVariable localVar = (FlowExpressions.LocalVariable) r;
+            CFValue current = localVariableValues.get(localVar);
+            CFValue value = changeLockAnnoToTop(r, current);
             localVariableValues.put(localVar, value);
         } else if (r instanceof FlowExpressions.FieldAccess) {
             FlowExpressions.FieldAccess fieldAcc = (FlowExpressions.FieldAccess) r;
+            CFValue current = fieldValues.get(fieldAcc);
+            CFValue value = changeLockAnnoToTop(r, current);
             fieldValues.put(fieldAcc, value);
         } else if (r instanceof FlowExpressions.MethodCall) {
             FlowExpressions.MethodCall method = (FlowExpressions.MethodCall) r;
+            CFValue current = methodValues.get(method);
+            CFValue value = changeLockAnnoToTop(r, current);
             methodValues.put(method, value);
         } else if (r instanceof FlowExpressions.ArrayAccess) {
             FlowExpressions.ArrayAccess arrayAccess = (ArrayAccess) r;
+            CFValue current = arrayValues.get(arrayAccess);
+            CFValue value = changeLockAnnoToTop(r, current);
             arrayValues.put(arrayAccess, value);
         } else if (r instanceof FlowExpressions.ThisReference) {
-            thisValue = value;
+            thisValue = changeLockAnnoToTop(r, thisValue);
         } else if (r instanceof FlowExpressions.ClassName) {
             FlowExpressions.ClassName className = (FlowExpressions.ClassName) r;
+            CFValue current = classValues.get(className);
+            CFValue value = changeLockAnnoToTop(r, current);
             classValues.put(className, value);
         } else {
             // No other types of expressions need to be stored.
         }
     }
 
+    /**
+     * Makes a new CFValue with the same annotations as currentValue except that the annotation
+     * in the LockPossiblyHeld hierarchy is set to LockPossiblyHeld. If currentValue is null,
+     * then a new value is created where the annotation set is LockPossiblyHeld and GuardedByUnknown
+     */
+    private CFValue changeLockAnnoToTop(Receiver r, CFValue currentValue) {
+        if (currentValue == null) {
+            currentValue =
```

Comment:
Instead of setting currentValue here, would it be nicer to set currentSet and currentType here, and then have shared code below? At the moment it looks a bit confusing when this new Value is used.